### DISCUSSION TITLE
Fast Data Draft Write - How can I know if my draft is still valid?

### DISCUSSION THREAD

**Discussion by frengooo** [mia-platform/community#354] [Ideas]
### Which area of this project could be most improved?

Fast Data

### Description

Hi guys,
I am talking with a prospect about using Fast Data for writing purposes.
I am not talking only about sending a "write command" to the System of Records, but about updating the data in _Draft_ mode on the Fast Data directly, in this way, I can expose two endpoints:
* GET SV &rarr; to extract the SVs with the data only from the _System of Records_;
* GET SV Draft &rarr; extract the SVs with draft data, if present.

It's not so easy because if the Fast Data allows the users to write in a _draft_ mode, the Fast Data needs to be sure that, when the same data is received from the _SoR_, the _draft_ data should be deleted.
Moreover, the Fast Data should allow to handle remediations, e.g. what happens if a "write command" is aborted by the _SoR_?

I was thinking about two possibilities to handle the _draft_ updates and the _draft_ deletion:
* using the Kafka messages that the Fast Data already have &rarr; if the CDC send a before-after, we can use it to check if a _draft_ data can be deleted, because the "write command" is completed. If the CDC doesn't send before-after, we can still use the "SV Updates" events produced by the Single Views Creator;
* using ksqlDB &rarr; it should be possible to use ksqlDB to query the ingestion topics to know if there is an updated version of the draft data and, if present, delete the draft.

To handle the "write commands" abort from the System of Records I have only two ideas at the moment:
* expose a REST API to explicitly abort a draft;
* set an expiration on draft fields that can be aborted by the system of record (e.g. a user changes its address but the SoR address verification process sets the address as "invalid").

What do you think about this?

Thanks 😄

**Comment by flower-of-the-bridges** [comment]
Hi @silversoul93 

I think you can try to achieve this use case by using the `__STATE__` property used by CRUD Collections. 
If you put the `__STATE__` field in the Single View Key, you can have two different documents in the same collection.

Then, you can expose the collection under one CRUD endpoint, where using the `_st=DRAFT` can enable you to look into draft records.

Then, by using [custom delete functions](https://docs.mia-platform.eu/docs/fast_data/configuration/single_view_creator/upsert_delete_strategies), you can delete documents accordingly on the message state, to have just one document once the draft operation has been completed.

**Reply by frengooo** [reply]
Hi, @flower-of-the-bridges, thanks for your response.
it should work, but just if you set the `__STATE__` as a Single View key as you told.

I am just confused about editing a Single View.
With the approach you suggested how can you create a draft version? Do you need to copy the "original" SV and modify some fields? Or just the SV Key and the draft data?

I like more the "incremental" approach, in which the "draft" piece is completely separated, but of course, it has pros and cons too.

**Reply by flower-of-the-bridges** [reply]
To create a draft version, you should have a dedicated single view creator plugin where you can use the [`onState` advanced property](https://docs.mia-platform.eu/docs/fast_data/configuration/config_maps/aggregation#aggregate-documents-with-different-__state__-other-than-public) of the aggregation configuration.

In this way, the main single view creator will only aggregate the public single views and perform the delete logic of draft records 

Using this approach, you can separate the two cases but insist on the same collection (by setting the same env variable `SINGLE_VIEWS_COLLECTION`)
